File: src/Enterprise/C#/PrototypeExample.cs
using System;
using System.Collections.Generic;

public interface IPrototype<out T>
{
    T Clone();
}

public sealed class ServiceProfile : IPrototype<ServiceProfile>
{
    public string Name { get; set; }
    public List<string> Features { get; }

    public ServiceProfile(string name, IEnumerable<string> features)
    {
        Name = name;
        Features = new List<string>(features);
    }

    private ServiceProfile(ServiceProfile source)
    {
        Name = source.Name;
        Features = new List<string>(source.Features);
    }

    public ServiceProfile Clone() => new(this);

    public string Describe() => $"{Name}: {string.Join(",", Features)}";
}

public static class PrototypeExample
{
    public static void Main()
    {
        IPrototype<ServiceProfile> prototype = new ServiceProfile("orders", new[] { "metrics" });
        var original = (ServiceProfile)prototype;
        var canary = prototype.Clone();

        canary.Name = "orders-canary";
        canary.Features.Add("tracing");

        Console.WriteLine($"original={original.Describe()}");
        Console.WriteLine($"clone={canary.Describe()}");
    }
}

File: src/Enterprise/Java/PrototypeExample.java
import java.util.ArrayList;
import java.util.List;

interface Prototype<T> {
    T copy();
}

final class ServiceProfile implements Prototype<ServiceProfile> {
    private String name;
    private final List<String> features;

    ServiceProfile(String name, List<String> features) {
        this.name = name;
        this.features = new ArrayList<>(features);
    }

    @Override
    public ServiceProfile copy() {
        return new ServiceProfile(name, features);
    }

    void setName(String name) {
        this.name = name;
    }

    void addFeature(String feature) {
        features.add(feature);
    }

    String describe() {
        return name + ": " + String.join(",", features);
    }
}

public final class PrototypeExample {
    public static void main(String[] args) {
        Prototype<ServiceProfile> prototype = new ServiceProfile("orders", List.of("metrics"));
        ServiceProfile original = (ServiceProfile) prototype;
        ServiceProfile canary = prototype.copy();

        canary.setName("orders-canary");
        canary.addFeature("tracing");

        System.out.println("original=" + original.describe());
        System.out.println("clone=" + canary.describe());
    }
}
